File: string/number-of-lines-to-write-string/number-of-lines-to-write-string.go
package numberoflinestowritestring

func numberOfLines(widths []int, s string) []int {
	buffer := 100
	lines := 0

	for i := 0; i < len(s); {
		try := buffer - widths[s[i]-'a']
		if try <= 0 {
			lines++
			buffer = 100
			if try == 0 {
				i++
			}
		} else {
			buffer = try
			i++
		}
	}

	last := buffer
	if buffer < 100 {
		lines++
		last = 100 - last
	}

	return []int{
		lines,
		last,
	}
}

File: string/number-of-lines-to-write-string/number-of-lines-to-write-string.ts
function numberOfLines(widths: number[], s: string): number[] {
    let buffer = 100;
    let lines = 0;

    for (let i = 0; i < s.length;) {
        const attempt = buffer - widths[s[i].charCodeAt(0) - 'a'.charCodeAt(0)];

        if (attempt < 0) {
            lines++;
            buffer = 100;
        } else {
            buffer = attempt;
            i++;
        }
    }

    if (buffer < 100) {
        lines++;
    }

    return [lines, 100 - buffer];
};
